tutorials/arvo/ames.md

```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+
+When a ship first contacts another ship, it may only know its `@p`, but an IP
+address and port are necessary for peer-to-peer communication to occur.
+Thus the initial correspondence between two ships, or additional correspondence
+following a change in IP or port of one of the parties, will need to be relayed
+by an intermediary that knows the IP and port of the receiving ship.
+
+For now packet relaying is handled entirely by galaxies. Every galaxy is
+responsible for knowing the IP and port of every planet underneath it. In the
+future, galaxies will only need to know the star sponsoring a planet, and stars
+will be responsible for the final step of the relay.
+
+The following diagram summarizes the packet creation and forwarding process.
+
+<div style="text-align:center">
+<img src="https://media.urbit.org/docs/arvo/datagram-long.png">
+</div>
+
+We elaborate on the above diagram.
+
+A typical relay will look something like this: `~bacbel-tagfeb` wishes to
+contact `~worwel-sipnum` but does not know the IP address and port at which
+`~worwel-sipnum` resides. Both planets live under `~zod`, which knows both of
+their IP and port numbers by virtue of being the galaxy that both live under.
+
+So `~bacbel-tagfeb` sends a packet addressed to `~worwel-sipnum` to `~zod`.
+
+The packet has the following format:
+ - The standard Ames header described in [header](#header), where the checksum
+   is the `+mug` of the body, and the sender and receiver ship types are `01`,
+   denoting that the sender and receiver are planets. The remainder of the
+   packet is the body.
+ - The body of this packet will be the `@p` of the
+sender `~bacbel-tagfeb`, followed by the `@p` of the receiver `~worwel-sipnum`,
+followed by the payload.
+ - The payload of this packet will be the `+jam` of `[origin content]`. `origin`
+for this initial packet will be `~`, which implies that the packet is
+originating from the sender. `content` is the `+jam` of the message, encrypted
+using a Diffie-Hellman symmetric key obtained with the private key of
+`~bacbel-tagfeb` and the public key of `~worwel-sipnum`.
+
+`~zod` receives the packet and reads the body. It sees that it is not the
+intended recipient of the packet, and so gets ready to forward it to
+`~worwel-sipnum`. First, it [`+cue`s](@/docs/reference/library/2p.md#cue)
+(deserializes) the payload and changes the `origin` to the IP and port of
+`~bacbel-tagfeb`. Then it `+jam`s `[origin content]` to form a new payload, and
+`+mug`s that payload to get a new checksum. It replaces the old payload and
+checksum with the new, and then sends the packet along to `~worwel-sipnum`.
+
+Once `~worwel-sipnum` processes the packet, it will know the IP and port of
+`~bacbel-tagfeb` since `~zod` included it when it forwarded the packet and so
+will send an ack packet directly to `~bacbel-tagfeb`. Communication between the
+two ships will now be direct until one of them changes their IP address, port, or networking keys.
```

Comment:
NATs and firewalls can prevent a direct peer-to-peer connection, even if the two ships know each other's IP and port.  In that case, a star or galaxy will continue to relay packets.